provider/packet/environ.go

```diff
@@ -0,0 +1,296 @@
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "")
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tag, state string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	opt := &packngo.ListOptions{Search: tag}
+
+	devices, _, err := e.packetClient.Devices.List(e.cloud.Credential.Attributes()["project-id"], opt)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range devices {
+		if state == "" || d.State == state {
+			toReturn = append(toReturn, &packetDevice{e, &d})
+		} 
+	}
+
+	return toReturn, nil
+}
+
+func (e *environ) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "active")
+}
+
+func (e *environ) Config() *config.Config {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfg.config
+}
+
+func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
+}
+
+func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
+	return nil, nil
+}
+
+func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
+	return nil
+}
+
+func (e *environ) Destroy(ctx context.ProviderCallContext) error {
+	return common.Destroy(e, ctx)
+}
+
+func (e *environ) DestroyController(ctx context.ProviderCallContext, controllerUUID string) error {
+	_, err := e.packetClient.Devices.Delete(controllerUUID, true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+func (e *environ) InstanceTypes(context.ProviderCallContext, constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
+	var i envinstance.InstanceTypesWithCostMetadata
+	return i, nil
+}
+
+func (e *environ) Instances(ctx context.ProviderCallContext, ids []instance.Id) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	for _, id := range ids {
+		//TODO handle case when some of the instanes are missing
+		d, _, err := e.packetClient.Devices.Get(string(id), nil)
+		if err != nil {
+			return nil, errors.Annotatef(err, "looking up device with ID %q", id)
+		}
+		toReturn = append(toReturn, &packetDevice{e, d})
+
+	}
+	if len(toReturn) == 0 {
+		return nil, environs.ErrNoInstances
+	}
+	return toReturn, nil
+}
+
+func (e *environ) MaintainInstance(ctx context.ProviderCallContext, args environs.StartInstanceParams) error {
```

Comment:
We don't need to implement this method. It was removed from the `InstanceBroker` indirection with #12151.